### PR TITLE
Introduce sparse matrices and required operations

### PR DESCRIPTION
**Pull request by DanielRrr**

### CODE REVIEW

**Review by arthurpaulino — CHANGES_REQUESTED** [review]

File: YatimaStdLib/SparseMatrix.lean [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+/--
+Creates a sparse matrix from a list
+-/
+def ofList (l : List ((Nat × Nat) × R)) : SparseMatrix R :=
+  let entries := Std.RBMap.ofList l compare
+  SparseMatrix.mk
+    entries
+    (1 + (Array.foldr max 0 $ Array.map (fun p => p.1) $ keysArray entries))
+    (1 + (Array.foldr max 0 $ Array.map (fun p => p.2) $ keysArray entries))
```
Comment (arthurpaulino) [inline YatimaStdLib/SparseMatrix.lean:38]: These should be parameters of the function. Suppose we have a 1000x1000 matrix whose only non-zero value (10) is at (0, 0). That would be `.ofList ((0, 0), 10) 1000 1000`.

Maybe `ofList` is not a good name. What about `ofValuesWithDims` or something similar?

Comment (DanielRrr) [reply YatimaStdLib/SparseMatrix.lean:38]: Makes sense

File: YatimaStdLib/SparseMatrix.lean [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+-/
+def SparseMatrix.addition
+  (m₁ : SparseMatrix R) (m₂ : SparseMatrix R) : SparseMatrix R := Id.run do
+  let cij i j := findD m₁.entries (i,j) 0 + findD m₂.entries (i,j) 0
+  let (cols, rows) := (m₁.cols, m₁.rows)
+  if dim m₁ == dim m₂ then
+    let mut acc : SparseMatrix R := SparseMatrix.empty cols rows
+    for i in [0:rows] do
+      for j in [0:cols] do
+        acc := SparseMatrix.insert i j (cij i j) acc
```
Comment (arthurpaulino) [inline YatimaStdLib/SparseMatrix.lean:125]: This logic will fill the sparse matrix with 0's, which is suboptimal. We want to have as little data as possible.

We need some kind of paired match with `find?`. For addition, the logic is:
- If we find elements on both matrices, we insert their sum
- If we only find one element and it's 0, we do nothing
- If we only find one element, insert it
- If we don't find any element, do nothing.

The general idea is that the resulting matrix will have no data other than non-zero elements

Comment (DanielRrr) [reply YatimaStdLib/SparseMatrix.lean:125]: Let me see

**Pull request merged**